packages/nitro-protocol/docs/force-move-safety.md

```diff
@@ -0,0 +1,399 @@
+| Chal(n,s,p) | checkpoint(m,S)    | Open(m)     | matches(S)   |
+|             |                    |             | m > n        |
+| Chal(n,s,p) | refute(m,s',S)     | Open(n)     | matches(S)   |
+|             |                    |             | m > n        |
+|             |                    |             | p == sig(s') |
+| Chal(n,s,p) | respond(s',S)      | Open(n+1)   | matches(S)   |
+|             |                    |             | s->s'        |
+
+This offers Alice a new strategy, applying `checkpoint(n, t, S)`, regardless of the current state.
+
+- Alice can call `checkpoint(n,t,S)` at any time with her most recent state,which has turn `n`
+- If this transaction gets recorded,it moves to `Open(n)`
+- The only valid actions from this point are
+  - `checkpoint(m,S') >> Open(n) == Open(m)` for some `m > n`
+  - `forceMove(m,s,p,S') >> Open(n) == Challenge(m,s,p)` for some `m > n`
+- Alice is happy
+
+## Attack
+
+- Suppose we're in `Open(0)`
+- Alice wants to apply `checkpoint(n,S)`
+- Bob front-runs with `forceMove(0,s,p)`
+
+```
+   checkpoint(n,Alice,S) >> forceMove(0,s',Eve,S) >> Open(0)
+== checkpoint(n,Alice,S) >> Challenge(0,s',Eve)
+== Challenge(0,s',Eve)
+```
+
+No matter how the challenge is cleared, Bob can repeat this, blocking the channel for time at least proportional to `n`.
+
+# Version 4
+
+By exposing the `turnNumRecord` using a technique inspired from [here](https://medium.com/@novablitz/storing-structs-is-costing-you-gas-774da988895e), the participant no longer has to provide the current channel state for the `forceMove` or `checkpoint` actions.
+
+| State       | Action           | NextState   | Requirements |
+| ----------- | ---------------- | ----------- | ------------ |
+| Open(n)     | forceMove(m,s,p) | Chal(m,s,p) | m >= n       |
+| Open(n,S)   | checkpoint(m)    | Open(m)     | m > n        |
+| Chal(n,s,p) | checkpoint(m)    | Open(m)     | m > n        |
+| Chal(n,s,p) | refute(m,s',S)   | Open(n)     | matches(S)   |
+|             |                  |             | m > n        |
+|             |                  |             | p == sig(s') |
+| Chal(n,s,p) | respond(s',S)    | Open(n+1)   | matches(S)   |
+|             |                  |             | s->s'        |
+
+In this version of the protocol, the `matches` function is implemented in Solidity as follows.
+
+```
+function matches(ChannelState memory S) public pure returns (bool) {
+  // Casting from uint256 to uint160 returns the 160 least significant bits.
+  // The 96 most significant bits can be used to store un-hashed data, including
+  // the turnNumRecord
+  return uint160(uint256(keccack256(S))) == uint160(uint256(currentState)) ;
+}
+```
+
+Thus, the `forceMove` and `checkpoint` actions are always enabled, if the participant can provide a newer state than the current `turnNumRecord`.
+This counters the attacks found on the previous versions, thanks to the commutativity of the `forceMove` and `checkpoint` actions:
```

Comment:
Perhaps include a subclause to remind the uninitiated about what "commutativity" means.